advanced/Scripts/piholeDebug.sh

```diff
@@ -9,534 +9,1071 @@
+  for i in "${!ports_in_use[@]}"; do
+    # loop through them and assign some local variables
+    local port_number
+    port_number="$(echo "${ports_in_use[$i]}" | awk '{print $1}')"
+    local service_name
+    service_name=$(echo "${ports_in_use[$i]}" | awk '{print $2}')
+    # Use a case statement to determine if the right services are using the right ports
+    case "${port_number}" in
+      53) compare_port_to_service_assigned  "${resolver}"
+          ;;
+      80) compare_port_to_service_assigned  "${web_server}"
+          ;;
+      4711) compare_port_to_service_assigned  "${ftl}"
+          ;;
+      # If it's not a default port that Pi-hole needs, just print it out for the user to see
+      *) log_write "[${port_number}] is in use by ${service_name}";
+    esac
+  done
 }
 
-daemon_check() {
-  # Check for daemon ${1} on port ${2}
-	header_write "Daemon Process Information"
+check_networking() {
+  # Runs through several of the functions made earlier; we just clump them
+  # together since they are all related to the networking aspect of things
+  echo_current_diagnostic "Networking"
+  detect_ip_addresses "4"
+  detect_ip_addresses "6"
+  ping_gateway "4"
+  ping_gateway "6"
+  check_required_ports
+}
 
-	echo ":::     Checking ${2} port for ${1} listener."
+check_x_headers() {
+  # The X-Headers allow us to determine from the command line if the Web
+  # server is operating correctly
+  echo_current_diagnostic "Dashboard and block page"
+  # Use curl -I to get the header and parse out just the X-Pi-hole one
+  local block_page
+  block_page=$(curl -Is localhost | awk '/X-Pi-hole/' | tr -d '\r')
+  # Do it for the dashboard as well, as the header is different than above
+  local dashboard
+  dashboard=$(curl -Is localhost/admin/ | awk '/X-Pi-hole/' | tr -d '\r')
+  # Store what the X-Header shoud be in variables for comparision later
+  local block_page_working
+  block_page_working="X-Pi-hole: A black hole for Internet advertisements."
+  local dashboard_working
+  dashboard_working="X-Pi-hole: The Pi-hole Web interface is working!"
+  # If the X-header found by curl matches what is should be,
+  if [[ $block_page == "$block_page_working" ]]; then
+    # display a success message
+    log_write "$TICK ${COL_LIGHT_GREEN}${block_page}${COL_NC}"
+  else
+    # Otherwise, show an error
+    log_write "$CROSS ${COL_LIGHT_RED}X-Header does not match or could not be retrieved.${COL_NC}"
+  fi
 
-	if [[ ${IPV6_READY} ]]; then
-	  port_check 6 "${2}" "${1}"
+  # Same logic applies to the dashbord as above, if the X-Header matches what a working system shoud have,
+  if [[ $dashboard == "$dashboard_working" ]]; then
+    # then we can show a success
+    log_write "$TICK ${COL_LIGHT_GREEN}${dashboard}${COL_NC}"
+  else
+    # Othewise, it's a failure since the X-Headers either don't exist or have been modified in some way
+    log_write "$CROSS ${COL_LIGHT_RED}X-Header does not match or could not be retrieved.${COL_NC}"
```

Comment:
There's some X headers set in the lighttpd config, and this checks if those headers are being added correctly. One is for the web interface, another is for anything else (most likely blocked ad)